### PR TITLE
GSB: Clean up resolveDeclSignature() calls

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: lib/AST/GenericSignatureBuilder.cpp [hunk 1]
```diff
@@ -2084,13 +2084,6 @@ TypeDecl *EquivalenceClass::lookupNestedType(
           foundMembers);
       for (auto member : foundMembers) {
         if (auto type = dyn_cast<TypeDecl>(member)) {
-          // Resolve the signature of this type.
```
Comment (slavapestov) [inline lib/AST/GenericSignatureBuilder.cpp]: @DougGregor there's another place where we do this, and if I remove it, all tests pass, except we no longer diagnose:

```
protocol P {
  typealias T = Self.T
}
```

I guess the next step is to start enabling request evaluator cycle diagnostics...

Comment (DougGregor) [reply lib/AST/GenericSignatureBuilder.cpp]: Yeah, I expect we'll learn a lot from experimenting with enabling request-evaluator cycle diagnostics.

File: lib/AST/GenericSignatureBuilder.cpp [hunk 2]
```diff
@@ -3812,25 +3805,31 @@ PotentialArchetype *GenericSignatureBuilder::realizePotentialArchetype(
   return pa;
 }
 
+static Type getStructuralType(TypeDecl *typeDecl, LazyResolver *resolver) {
```
Comment (DougGregor) [inline lib/AST/GenericSignatureBuilder.cpp]: I'd drop the `LazyResolver` parameter here. We can get it from `typeDecl->getASTContext().getLazyResolver()` in the few cases where we need it.

File: lib/AST/GenericSignatureBuilder.cpp [hunk 3]
```diff
@@ -4226,34 +4225,13 @@ ConstraintResult GenericSignatureBuilder::expandConformanceRequirement(
     return result;
   };
 
-  // Form an unsubstituted type referring to the given type declaration,
-  // for use in an inferred same-type requirement.
-  auto formUnsubstitutedType = [&](TypeDecl *typeDecl) -> Type {
```
Comment (DougGregor) [inline lib/AST/GenericSignatureBuilder.cpp]: Well, that's nice to delete. Bravo!

File: lib/AST/GenericSignatureBuilder.cpp (hunk 2 above)
Comment (slavapestov) [reply lib/AST/GenericSignatureBuilder.cpp]: I think @xymus can do that next, after landing the getStructuralType() request :)

**Pull request merged**